docs/kernel_providers.rst

```diff
@@ -0,0 +1,173 @@
+            if not which('oblong-kernel'):
+                return  # Check it's available
+
+            # Two variants - for a real kernel, these could be something like
+            # different conda environments.
+            yield 'standard', {
+                'display_name': 'Oblong (standard)',
+                'language': {'name': 'oblong'},
+                'argv': ['oblong-kernel'],
+            }
+            yield 'rounded', {
+                'display_name': 'Oblong (rounded)',
+                'language': {'name': 'oblong'},
+                'argv': ['oblong-kernel'],
+            }
+
+        def launch(self, name, cwd=None, launch_params=None):
+            if name == 'standard':
+                return my_launch_method(cwd=cwd, launch_params=launch_params,
+                                        kernel_cmd=['oblong-kernel'],
+                                        extra_env={'ROUNDED': '0'})
+            elif name == 'rounded':
+                return my_launch_method(cwd=cwd, launch_params=launch_params,
+                                        kernel_cmd=['oblong-kernel'],
+                                        extra_env={'ROUNDED': '1'})
+            else:
+                raise ValueError("Unknown kernel %s" % name)
+
+You would then register this with an *entry point*. In your ``setup.py``, put
+something like this::
+
+    setup(...
+        entry_points = {
+        'jupyter_kernel_mgmt.kernel_type_providers' : [
+            # The name before the '=' should match the id attribute
+            'oblong = oblong_provider:OblongKernelProvider',
+        ]
+    })
+
+Finding kernel types
+====================
+
+To find and start kernels in client code, use
+:class:`KernelFinder <jupyter_kernel_mgmt.discovery.KernelFinder>`. This uses multiple kernel
+providers to find available kernels. Like a kernel provider, it has methods
+``find_kernels`` and ``launch``. The kernel names it works
+with have the provider ID as a prefix, e.g. ``oblong/rounded`` (from the example
+above).
+
+::
+
+    from jupyter_kernel_mgmt.discovery import KernelFinder
+    kf = KernelFinder.from_entrypoints()
+
+    ## Find available kernel types
+    for name, attributes in kf.find_kernels():
+        print(name, ':', attributes['display_name'])
+    # oblong/standard : Oblong (standard)
+    # oblong/rounded : Oblong(rounded)
+    # ...
+
+    ## Start a kernel by name
+    connect_info, manager = kf.launch('oblong/standard')
+
+    # TODO - Finish example with IOLoopKernelClient, etc.
+
+Included kernel providers
+=========================
+
+``jupyter_kernel_mgmt`` includes two kernel providers in its distribution.
+
+1. :class:`KernelSpecProvider <jupyter_kernel_mgmt.discovery.KernelSpecProvider>` handles the discovery and launch
+of most existing kernelspec-based kernels that exist today.  If a provider id is not provided
+when calling :meth:`KernelFinder.launch(name) <.KernelFinder.launch>`, KernelFinder will assume its associated with
```

Comment:
KernelFinder doesn't do this - it's up to the application to do so. Open for debate, but I think it makes more sense to do it at the application layer.